core/contacts.go

```diff
@@ -136,95 +164,119 @@ func (t *Textile) UpdateContactInboxes() error {
 
-	// search the network via cafe if present
-	var inbound []*pb.Contact
-	sessions := t.datastore.CafeSessions().List()
-	if len(sessions) == 0 {
-
-		// search via pubsub directly
-		res, err := t.cafe.FindContactPubSub(&pb.CafeContactQuery{
-			FindId:       query.Id,
-			FindAddress:  query.Address,
-			FindUsername: query.Username,
-			Limit:        int32(query.Limit),
-			Wait:         int32(query.Wait),
-		}, false)
-		if err != nil {
-			return nil, err
-		}
-		inbound = res.Contacts
+	go func() {
+		defer func() {
+			for _, ch := range searchChs {
+				close(ch)
+			}
+		}()
 
-	} else {
-		groups := make([][]*pb.Contact, len(sessions))
-		wg := sync.WaitGroup{}
-		var ferr error
-		for i, session := range sessions {
-			cafe, err := peer.IDB58Decode(session.Id)
-			if err != nil {
-				return result, err
+		// search local
+		for _, c := range t.datastore.Contacts().Find(query.Id, query.Address, query.Username) {
+			added := set.Add(repoContactToProto(&c))
+			if len(added) == 0 {
+				continue
+			}
+			info := t.contactInfo(&c, true)
+			if info != nil {
+				localCh <- &ContactQueryResult{Local: true, Contact: *info}
 			}
-			wg.Add(1)
-			go func(i int, cafe peer.ID) {
-				defer wg.Done()
-				res, err := t.cafe.FindContact(query, cafe)
-				if err != nil {
-					ferr = err
-					return
-				}
-				groups[i] = res
-			}(i, cafe)
-		}
-		wg.Wait()
-		if ferr != nil {
-			return result, ferr
 		}
 
-		for _, g := range groups {
-			inbound = deduplicateContactResults(append(inbound, g...))
+		if query.Local || len(set.items) >= query.Limit {
+			return
 		}
-	}
 
-	for _, i := range inbound {
-		contact := protoContactToRepo(i)
-		if query.Add {
-			ex := t.datastore.Contacts().Get(contact.Id)
-			if ex == nil || ex.Updated.UnixNano() < contact.Updated.UnixNano() {
-				if err := t.datastore.Contacts().AddOrUpdate(contact); err != nil {
-					return nil, err
+		// search the network via cafe if present
+		if len(sessions) == 0 {
+
+			// search via pubsub directly
+			canceler := cancel.Listen()
+			defer canceler.Close()
+			if err := t.cafe.FindContactPubSub(&pb.CafeContactQuery{
+				FindId:       query.Id,
+				FindAddress:  query.Address,
+				FindUsername: query.Username,
+				Limit:        int32(query.Limit),
+				Wait:         int32(query.Wait),
+			}, set, func(res *pb.CafeContactQueryResult) {
+				for _, c := range res.Contacts {
+					contact := t.contactInfo(protoContactToRepo(c), false)
+					if contact != nil {
+						clientCh <- &ContactQueryResult{Contact: *contact}
+					}
 				}
+			}, canceler.Ch, false); err != nil {
+				errCh <- err
+				return
 			}
-		}
 
-		info := t.contactInfo(contact, false)
-		if info != nil {
-			result.Remote = append(result.Remote, *info)
+		} else {
+
+			// search via cafes
+			wg := sync.WaitGroup{}
+			for i, session := range sessions {
+				cafe, err := peer.IDB58Decode(session.Id)
+				if err != nil {
+					errCh <- err
+					return
+				}
+				canceler := cancel.Listen()
```

Comment:
For each concurrent search, create a broadcast listener that can be invoked by a single caller.